### ISSUE TITLE
AsyncSchemaRegistryDeserializer creates new delegate for each invocation of any deserializer

### ISSUE THREAD

**Issue by fabianoliver** [ch-robinson/dotnet-avro#314]
I was going through a memory dump and noticed unusually high allocations coming from calls to `AsyncSchemaRegistryDeserializer<T>.DeserializeAsync` on a process running on Win10 under .NET8. 

The allocated types were of type `Delegate`, with their call stack going `DeserializeAsync` -> `System.Reflection.Emit.DynamicMethod.CreateDelegate(Type, Object)` -> `Delegate.CreateDelegatenoSecurityCheck`.

It appears that every call to `DeserializeAsync` leads to a new delegate being created for every single deserialization call (!!).
For example, set breakpoint in Delegate.CoreCLR.cs -> `CreateDelegateNoSecurityCheck`, and run something like 

```
  var s = new AsyncSchemaRegistryDeserializer<Something>(...);
  await s.DeserializeAsync(somePayload, false, default);
  await s.DeserializeAsync(somePayload, false, default);
  await s.DeserializeAsync(somePayload, false, default);
```

`CreateDelegateNoSecurityCheck` will be hit (at least?) three times.

I haven't had much time to dig in in greater detail why that is happening, my suspicion is the following though: [line 277](https://github.com/ch-robinson/dotnet-avro/blob/main/src/Chr.Avro.Confluent/Confluent/AsyncSchemaRegistryDeserializer.cs#L277), if we look at 

```
Expression.Lambda<Func<ReadOnlyMemory<byte>, T>>(reader, memory).Compile().GetMethodInfo().GetParmaeters()
```

The compiled method actually has two parameters: The latter one being `ReadOnlyMemory<byte>`, but the former being of type `System.Runtime.CompilerServices.Closure`. This looks a bit like somehow the compiled lambda is capturing closure, which I imagine might lead to re-creation of the delegates each call - albeit I haven't been able to confirm this with certainty yet.

**Comment by dstelljes** [comment]
Hmm, I would guess that this is because inner `LambdaExpression`s aren't compiled when the outer expression is compiled. Chr.Avro generates a `LambdaExpression` for each record type to support recursive references. (To confirm, you should be able to verify that (de)serializers for non-record types do not call `CreateDelegate`.)

Given this, we might have to revisit how we build expression trees.

**Comment by fabianoliver** [comment]
I'm still trying to understand what .NET is doing under the hood there. I don't think it's an issue of _compilation_ per se - in the sense that the nested lamda is compiled fine I think.
I'm not even sure it'd really count as a proper bug, whether delegates are or are not cached is definitely more art than science in C# I think, with the how and ifs even having changed over [language versions](https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-11#improved-method-group-conversion-to-delegate). So probably "just" more of a performance downside particularly noticeable in scenarios with fairly high message rates.

As for delegate creation, my understanding was that non-capturing lambdas are cached,  but that doesn't seem to be happening for expressions (though I'd need to double check if the language specs say they _can_ be cached, or _must_ be cached).

The simplest example I could condense it down to so far is the following:
```
var param = Expression.Parameter(typeof(int));
var innerParam = Expression.Parameter(typeof(Action<int>));
var inner = Expression.Lambda(innerParam.Type, Expression.Block(), "test", new[] { param });
var action = Expression.Lambda<Action<int>>(Expression.Block(new[] { innerParam }, Expression.Assign(innerParam, inner), Expression.Invoke(innerParam, param)), param).Compile();

// Each invocation will trigger delegate creation
action(456);
action(567);
```

Interestingly, I don't observe the same behaviour when creating an equivalent function/invocation structure directly in code outside of the expression system:

```
Action<int> inner = _ => { };
Action<int> action= x =>
{
  Action<int> innerParam = inner;
  innerParam(x);
};

// Delegates are cached just fine
action(456);
action(567);
```

Even more interestingly, [FastExpressionCompiler](https://github.com/dadhi/FastExpressionCompiler) seems to be doing a better job here. If I compile any of my samples with that, delegates do seem to be cached fine - though I may have to run this through for more complex examples.

**Comment by fabianoliver** [comment]
After a bit more testing - FastExpressionCompiler isn't able to compile more complex lambda expressions / throws in those cases, so unfortunately doesn't seem to be a great workaround.

Thinking about this a little further, I _suspect_ that in the current way expression trees are created, re-creating delegates might be unavoidable. Not in the simple examples above, but at least once you have types whose properties are other complex types (whether self-referential or not) - if I understand the current generators right, their lambda would reference variables from the outer scope to access their nested types' deserializers. I think that would inevitably make the lambda capture the outer parameter, and capturing lambdas are never cached. Though hard to verify at the moment given even the simple, non-capturing versions aren't cached in the default compilation path.

I think a few options in increasing order of madness that come to mind at the moment are:
- Keep everything as-is. Which might be a perfectly sane choice if the alternatives are too complex / not worth the gain in performance and reduced allocs!!
- Try to wire up circular references in the expression tree in another way. An interesting technique for anonymous recursion is described [here](https://learn.microsoft.com/de-de/archive/blogs/wesdyer/anonymous-recursion-in-c) for example - maybe it'd similarly be able to create some variation of `delegate object DeserializeRecord(Type dtoType, string recordName, ref BinaryReader reader)`, then have each generated record deserializer take a delegate of that signature that'd allow it to deserialize nested types without having to capture/reference something from the outer scope? Hard to say for sure though if that'd really do the trick
- Instead of an expression tree, explicitly create an AnonymousType with one AnonymousMethod to deserialize each record type. Emitting IL directly sounds like a nightmare though and I'm not sure if there's much in terms of supporting libraries out there that could make it a bit easier.
- Ditching all that; generate "proper" full source for a deserializer, and use Roslyn to compile and load up the assembly. It wouldn't get more flexible than that, and leaving behind "almost-abandoned-by-microsoft" expression trees could be nice as well. However.. pulling in those giant dependencies, plus dealing with possibly breaking changes (eg definite no for AOT or mobile devices; though not quite sure if expression trees work there either) probably makes this a non-starter anyway

Maybe the first option isn't so bad after all.. :-)

**Comment by dstelljes** [comment]
Thanks for the detailed writeup! A few additional thoughts:

> Possibly do a lazy workaround that'd generate a simpler expression tree in case where the record is a simple type that contains only primitive fields.

Of the options you laid out, something like that is probably the most pragmatic. (A more sophisticated variation might be to walk the type graph and only do the circular reference trick for types that appear in a cycle.)

> Maybe the first option isn't so bad after all.. :-)

This is more or less where we're at too. I'm going to leave this as a bug since it's an unintended consequence of how the serdes are built, but it's unlikely that we'll be able to prioritize working on it.

**Comment by fabianoliver** [comment]
Hi,

I've recently spent a bit of time optimizing an internal application of ours. After several iteration cycles of tuning, eventually this issue here, as well as [this issue](https://github.com/ch-robinson/dotnet-avro/issues/342) became the new bottlenecks of the app. Out of curiosity, I implemented slightly hacky workaround for both to scope out what the impact would be:
- Getting rid of a variation of the second issue (unmapped record types ultimately got deserialized into dynamo objects), throughput of our internal stress test increased by 32%
- Working around this issue here ("inline deserialization" of record types instead of an extra hop through a function) improved throughout by a further 70%

Fairly significant improvements there, in particular this wasn't an avro-specific microbenchmark, but an improvement of the overall system in a real world-ish stress test. 

Long story short, I'll probably turn the hacky workarounds into something more proper soon. Would you be interested in getting these changes into chr.avro here? If so, I'd be happy to try whipping up PRs (hopefull) in the coming weeks here.

I think what you suggested here:

> (A more sophisticated variation might be to walk the type graph and only do the circular reference trick for types that appear in a cycle.)

Would make a lot of sense, so this would probably the direction I'd try to take if so?
